### PR TITLE
Add libsystem_networkextension

### PR DESCRIPTION
**Pull request by CuriousTommy**
This is needed for Heimdal. The arguments for `NEHelperCacheCopyAppUUIDMapping ` and `ne_session_set_socket_attributes ` were guesses.

### CODE REVIEW

**Comment by bugaevc** [comment]
The folder should be just `src/networkextension`

File: src/libsystem_networkextension/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,27 @@
+project(system_networkextension)
+
+cmake_minimum_required(VERSION 2.4.0)
+
+add_definitions(-DTARGET_OS_MAC=1)
+add_definitions(-D__APPLE__ -D__DYNAMIC__)
+
+set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -nostdinc -D__DARWIN_UNIX03 -fPIC -w")
```
Comment (bugaevc) [inline src/libsystem_networkextension/CMakeLists.txt]: Aren't these flags and definitions already set for all projects in main CMakeLists?

File: src/libsystem_networkextension/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+set(DYLIB_INSTALL_NAME "/usr/lib/system/libsystem_networkextension.dylib")
+add_darling_library(system_networkextension
+    SHARED
+        ne_session.c
+        NEHelperClient.c
+)
+target_link_libraries(system_networkextension
+    PRIVATE
+        system_c
+        system_dyld
```
Comment (bugaevc) [inline src/libsystem_networkextension/CMakeLists.txt]: Should it not be circular? Nobody's going to link against it directly.

Comment (bugaevc) [reply src/libsystem_networkextension/CMakeLists.txt]: Probably not circular, just reexported

File: src/libsystem_networkextension/CMakeLists.txt (hunk 1 above)
Comment (CuriousTommy) [reply src/libsystem_networkextension/CMakeLists.txt]: Most of the stuff is copied from `libsystem_coreservices`'s CMakeLists.txt. I'll remove it.

File: src/libsystem/CMakeLists.txt [hunk 3]
```diff
@@ -75,6 +75,7 @@ target_link_libraries(system PRIVATE
 	corecrypto
 	commonCrypto
 	system_dnssd
+	system_networkextension
```
Comment (CuriousTommy) [inline src/libsystem/CMakeLists.txt]: I don't understand why but for some reason, darling is not able to find `/usr/lib/system/libsystem_networkextension.dylib`.

**Comment by CuriousTommy** [comment]
@bugaevc Once I fix everything else, I'll rename the folder.

File: src/libsystem_networkextension/CMakeLists.txt (hunk 1 above)
Comment (CuriousTommy) [reply src/libsystem_networkextension/CMakeLists.txt]: Removed it.

File: src/libsystem_networkextension/CMakeLists.txt (hunk 2 above)
Comment (CuriousTommy) [reply src/libsystem_networkextension/CMakeLists.txt]: Dumb question, but what would `libsystem_networkextension` reexport?

File: src/libsystem/CMakeLists.txt (hunk 3 above)
Comment (LubosD) [reply src/libsystem/CMakeLists.txt]: Look into cmake/use_ld64.cmake

File: src/libsystem_networkextension/CMakeLists.txt (hunk 2 above)
Comment (LubosD) [reply src/libsystem_networkextension/CMakeLists.txt]: Nothing. Only libSystem will re-export this library.

File: src/libsystem/CMakeLists.txt (hunk 3 above)
Comment (CuriousTommy) [reply src/libsystem/CMakeLists.txt]: That was the issue, thanks!

**Comment by CuriousTommy** [comment]
I think that should be everything, I'll rebase and merge if everyone is happy with the changes.

File: src/libsystem_networkextension/CMakeLists.txt (hunk 1 above)
Comment (bugaevc) [reply src/libsystem_networkextension/CMakeLists.txt]: > Most of the stuff is copied from libsystem_coreservices's CMakeLists.txt.

Yeah, we should clean up all the other subprojects of specifying common flags and options and includes too. Want to try doing that? :slightly_smiling_face:

**Pull request merged**